lib/rainforest/cli/uploader.rb

```diff
@@ -0,0 +1,162 @@
+# frozen_string_literal: true
+class RainforestCli::Uploader
+  attr_reader :test_files
+
+  def initialize(options)
+    ::Rainforest.api_key = options.token
+    @test_files = RainforestCli::TestFiles.new(options.test_folder)
+  end
+
+  # NOTE: Embedded tests must be successfully uploaded before parent tests.
+  def upload
+    validate_embedded_tests!
+
+    # Prioritize tests so that no parent tests are not uploaded before their
+    # children exist.
+    upload_groups = make_upload_groups
+
+    RainforestCli.logger.info 'Uploading tests...'
+
+    # Upload tests in parallel if there is only one upload group (no priority).
+    if upload_groups.count == 1
+      upload_group_in_parallel(upload_groups.first)
+    else
+      upload_groups_sequentially(upload_groups)
+    end
+  end
+
+  private
+
+  def validate_embedded_tests!
+    contains_nonexistent_ids = rfml_tests.select { |t| (t.embedded_ids - all_rfml_ids).any? }
+
+    if contains_nonexistent_ids.any?
+      raise TestsNotFound.new(contains_nonexistent_ids.map(&:file_name))
+    end
+  end
+
+  # Prioritize embedded tests before other tests
+  def make_upload_groups
```

Comment:
I still just don't quite understand what this is doing. Is it depth first search for embedded tests?